src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -250,14 +250,17 @@ private static PackageSpec ToPackageSpec(ProjectNames projectNames, IVsProjectRe
                         Path.Combine(
                             projectDirectory,
                             projectRestoreInfo.BaseIntermediatePath));
+
+            var projectName = GetPackageId(projectNames, projectRestoreInfo.TargetFrameworks);
+
             var packageSpec = new PackageSpec(tfis)
             {
-                Name = GetPackageId(projectNames, projectRestoreInfo.TargetFrameworks),
+                Name = projectName,
                 Version = GetPackageVersion(projectRestoreInfo.TargetFrameworks),
                 FilePath = projectFullPath,
                 RestoreMetadata = new ProjectRestoreMetadata
                 {
-                    ProjectName = projectNames.ShortName,
+                    ProjectName = projectName,
```

Comment:
Does this also apply to other project types like csproj PackageReference?

Reply:
It should. 
Add tests for Legacy as well.